### PR TITLE
Added Reducible[_] type class.

### PR DESCRIPTION
**Pull request by non**
This commit does a lot of related things. Here's a list:

 1. Removes existing `foldRight`, renames `foldLazy` to `foldRight`
 2. Makes `Fold[_]` instances a bit more powerful to work with
 3. Removes `Comonad[OneAnd[?, F]]` instance.
 4. Add some documentation.
 5. Some whitespace and formatting changes in a few files.

These changes are all possibly contentious.

(1) was intended to address the confusing bestiary of various
fold methods: `foldLeft`, `foldRight`, and `foldLazy`. Scala's eager
`foldRight` method is mostly-useless, which is why it seemed
reasonable to just remove it and replace it with our version.
I would be open to a different name -- I became convinced
`foldLazy` was a bad name but maybe I'm wrong?

(2) adds the `.imap` and `.complete` methods to `Fold[_]`. In
particular, .complete is something that almost anyone working
with `Fold[_]` will probably want.

(3) addresses a problem I noticed awhile ago: the `Comonad`
instance we had for `OneAnd` only dealt with the `.head`, dropping
the `.tail` on the floor. This is definitely wrong, although
we don't have a law to encode this yet. I noticed it while
working to have `OneAnd` implement `Reducible`.

(4) just adds a bit of documentation that I felt was missing.

(5) is arguably bad, but I tried to clean up some files that
I happened to be editing. I am torn between wanting to leave
lines alone (so that git blame is more useful and to preserve
the original author's intentions) and standardizing on a more
approachable style.

### CODE REVIEW

File: data/src/main/scala/cats/data/Or.scala [hunk 1]
```diff
@@ -0,0 +1,186 @@
+package cats
```
Comment (ceedubs) [inline data/src/main/scala/cats/data/Or.scala]: I think you might have accidentally added this back to git. It has been replaced by `Xor`.

**Comment by ceedubs** [comment]
I left a comment about naming on [the issue](https://github.com/non/cats/issues/145#issuecomment-95637877). I'm okay with this being named `Reducible`, but I thought I would just bring up the thoughts I had.

At some point, scalaz added `foldMapRight1` (which would probably be `reduceMapRight` for Cats) as an abstract method on `Foldable1`. This allows for optimizations in certain cases like [`Foldable1.toNel`](https://github.com/scalaz/scalaz/pull/824) and may be worth considering. @non is quite good about optimization, so it's possible he is already doing something efficient that I'm overlooking.

I think `UnsafeFoldable` should be documented/justified since it's unsafe.

There's a lot to digest in this PR and I'm not sure I absorbed all of it. Other than `Or.scala` being added back in, and `UnsafeFoldable` being documented/justified, I think I'm okay with this getting merged and possible future refinements coming later. Also a rebase is needed.

**Comment by non** [comment]
@ceedubs Good catch with `Or` -- I'll fix it.

I'll take a look at what Scalaz is doing. In principle this should be doing the same thing but I want to be sure.

**Comment by non** [comment]
@ceedubs I took your comments to heart. `reduceLeftTo` and `reduceRightTo` are the methods that would do this kind of work. I also removed the *unsafe* classes.

File: core/src/main/scala/cats/Foldable.scala [hunk 2]
```diff
@@ -132,17 +137,44 @@ import simulacrum._
...
+    foldLeft(fa, mutable.ListBuffer.empty[A]) { (buf, a) =>
+      if (p(a)) buf += a else buf
+    }.toList
+
+
+  /**
+   * Convert F[A] to a List[A], dropping all initial elements which
+   * match `p`.
+   */
+  def dropWhile_[A](fa: F[A])(p: A => Boolean): List[A] =
```
Comment (ceedubs) [inline core/src/main/scala/cats/Foldable.scala]: What's the reasoning/convention behind the underscores here and in `filter_`?

File: core/src/main/scala/cats/Reducible.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+  /**
+   * Overriden from Foldable[_] for efficiency.
+   */
+  override def reduceLeftToOption[A, B](fa: F[A])(f: A => B)(g: (B, A) => B): Option[B] =
+    Some(reduceLeftTo(fa)(f)(g))
+
+  /**
+   * Apply f to each element of fa and lazily combine them with the
+   * given associative function `g`.
```
Comment (ceedubs) [inline core/src/main/scala/cats/Reducible.scala]: Here and in `reduceLeftTo`, the ScalaDoc says "Apply f to each element", but wouldn't it only be applied to the initial element?

File: core/src/main/scala/cats/Reducible.scala [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  override def reduceRightToOption[A, B](fa: F[A])(f: A => B)(g: A => Fold[B]): Lazy[Option[B]] =
+    Lazy(Some(reduceRightTo(fa)(f)(g).value))
+
+  /**
+   * Traverse `F[A]` using `Apply[G]`.
+   *
+   * `A` values will be mapped into `G[B]` and combined using
+   * `Applicative#map2`.
+   *
+   * This method does the same thing as `Foldable#traverse_`.  The
```
Comment (ceedubs) [inline core/src/main/scala/cats/Reducible.scala]: Thanks. This is one of many really helpful comments in this PR.

File: core/src/main/scala/cats/data/OneAnd.scala [hunk 5]
```diff
@@ -12,108 +13,125 @@ package data
...
-
-    override def partialFold[A, B](fa: OneAnd[A,F])(f: A => Fold[B]): Fold[B] = {
-      import Fold._
-      f(fa.head) match {
-        case b @ Return(_) => b
-        case Continue(c) => foldable.partialFold(fa.tail)(f) match {
-          case Return(b) => Return(c(b))
-          case Continue(cc) => Continue { b => c(cc(b)) }
-          case _ => Continue(c)
+  implicit def oneAndFoldable[F[_]](implicit foldable: Foldable[F]): Foldable[OneAnd[?,F]] =
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/OneAnd.scala]: I think it should wait for another PR as this one is already pretty large, but it looks like we are missing a `oneAndReducible` instance. Is that right?

File: core/src/main/scala/cats/syntax/reducible.scala [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+      val typeClassInstance = U.TC
+    }
+}
+
+trait ReducibleSyntax extends Reducible.ToReducibleOps with ReducibleSyntax1 {
+  implicit def nestedReducibleSyntax[F[_]: Reducible, G[_], A](fga: F[G[A]]): NestedReducibleOps[F, G, A] =
+    new NestedReducibleOps[F, G, A](fga)
+}
+
+class NestedReducibleOps[F[_], G[_], A](fga: F[G[A]])(implicit F: Reducible[F]) {
```
Comment (ceedubs) [inline core/src/main/scala/cats/syntax/reducible.scala]: Is this intentionally left open for extension? I tend to default to `final` on classes if I don't really expect people to extend them, but I may go overboard.

File: tests/src/test/scala/cats/tests/CokleisliTests.scala [hunk 7]
```diff
@@ -19,18 +19,4 @@ class CokleisliTests extends CatsSuite {
...
-    type CokleisliNEL[A, B] = Cokleisli[NonEmptyList, A, B]
-
-    implicit def ev0[A: Arbitrary, B: Arbitrary]: Arbitrary[CokleisliNEL[A, B]] =
-      cokleisliArbitrary
-
-    implicit def ev1[A: Arbitrary, B: Eq]: Eq[CokleisliNEL[A, B]] =
-      cokleisliEq[NonEmptyList, A, B](oneAndArbitrary, Eq[B])
-
-    checkAll("Cokleisli[NonEmptyList, Int, Int]", ArrowTests[CokleisliNEL].arrow[Int, Int, Int, Int, Int, Int])
-    checkAll("Arrow[Cokleisli[NonEmptyList, ?, ?]]", SerializableTests.serializable(Arrow[CokleisliNEL]))
```
Comment (ceedubs) [inline tests/src/test/scala/cats/tests/CokleisliTests.scala]: Did we lose a couple tests here?

**Comment by ceedubs** [comment]
@non sorry it took me so long to get around to a review. I think this PR does a lot of great things, and I like the approaches it takes. I've left some questions, but once they are addressed, I plan to give this a thumbs-up.

File: core/src/main/scala/cats/Foldable.scala (hunk 2 above)
Comment (non) [inline core/src/main/scala/cats/Foldable.scala]: So, I took the naming convention from other places in `Foldable`, where you have less powerful versions of *sequence* and *traverse* which are named *sequence_* and *traverse_* (as opposed to the methods named in `Traversable`).

Similarly, in this case we can't *filter* or *dropWhile* to an `F[A]` value, but only to a `List[A]` (unlike something like `MonadFilter`). So I decided to use the same convention.

File: core/src/main/scala/cats/Reducible.scala (hunk 3 above)
Comment (non) [inline core/src/main/scala/cats/Reducible.scala]: Yes! Good catch, I'll fix it.

File: core/src/main/scala/cats/syntax/reducible.scala (hunk 6 above)
Comment (non) [inline core/src/main/scala/cats/syntax/reducible.scala]: I normally do like to mark things `final` too. I'll go ahead and do it here.

File: core/src/main/scala/cats/Foldable.scala (hunk 2 above)
Comment (ceedubs) [inline core/src/main/scala/cats/Foldable.scala]: Okay, that makes sense. If I have this question, I imagine others will too. I wonder if there's a good way to document it and make it easy to discover.

File: tests/src/test/scala/cats/tests/CokleisliTests.scala (hunk 7 above)
Comment (non) [inline tests/src/test/scala/cats/tests/CokleisliTests.scala]: These are no longer valid.

We need an `Arrow[CokleisliNEL]`.
This is equivalent to wanting `Arrow[Cokleisli[NonEmptyList, ?, ?]]`,
e.g. `Arrow[Cokleisli[OneAnd[?, List], ?, ?]]`.

The implicit that would be used here is:

```scala
  implicit def cokleisliArrow[F[_]](implicit ev: Comonad[F]): Arrow[Cokleisli[F, ?, ?]] =
    new CokleisliArrow[F] { def F: Comonad[F] = ev }
```

This requires a `Comonad[F]`, e.g. a `Comonad[OneAnd[?, List]]`.

However, the previous comonad instance for `OneAnd` was invalid, so I removed it.

Maybe I should implement an instance specifically for `NonEmptyList`?

Comment (ceedubs) [inline tests/src/test/scala/cats/tests/CokleisliTests.scala]: Oh I see. And I see that you have implemented the `Comonad` instance for `NonEmptyList`. Thanks!

**Comment by ceedubs** [comment]
:+1:

**Comment by stew** [comment]
:+1:

**Pull request merged**